### PR TITLE
Use correct simplifications after creating ifxcmpne

### PR DESCRIPTION
**Pull request by kevindean12**
Creates ifxcmpne helper methods in the simplifier (taken from and now called by the body of the corresponding ifxcmpneSimplifier() function) and calls them after calling simplifyISelectCompare() from an ifxcmpxx simplifier when the iselect compare simplification causes the IL op of the ifxcmpxx node to become an ifxcmpne, making the rest of the handler's logic potentially incorrect due to the assumption that the op being simplified is still the same ifxcmpxx as it had been.

Corrects reversed logic observed in cases where an ifxcmpxx simplifier continued simplifying after simplifyISelectCompare() had caused the node to become an ifxcmpne. For example, an ificmpeq was simplified and transformed into an ificmpne during simplifyISelectCompare(), and then when the ificmpeqSimplifier() continued simplifying after simplifyISelectCompare(), the result was reversed logic producing incorrectly compiled code.

### CODE REVIEW

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 1]
```diff
@@ -13037,166 +13044,121 @@ TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
    return node;
    }
 
-//---------------------------------------------------------------------
-// Integer if compare equal (signed and unsigned)
-//
-
-TR::Node *ificmpeqSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+static TR::Node*simplifyIficmpneHelper(TR::Node *node, TR::Block *block, TR::Simplifier *s)
```
Comment (0xdaryl) [inline compiler/optimizer/OMRSimplifierHandlers.cpp:13047]: Please add a space before the first `*`.

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 2]
```diff
@@ -13037,166 +13044,121 @@ TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
...
 
    TR::Node * firstChild = node->getFirstChild(), * secondChild = node->getSecondChild();
 
    if (firstChild == secondChild)
       {
-      s->conditionalToUnconditional(node, block, false);
+      if (!opChangedToCmpNE)
+         s->conditionalToUnconditional(node, block, true);
+      else
+         s->conditionalToUnconditional(node, block, false);
```
Comment (0xdaryl) [inline compiler/optimizer/OMRSimplifierHandlers.cpp:13147]: Rewrite these 4 lines as:
```
s->conditionalToUnconditional(node, block, !opChangedToCmpNE);
```

**Review by 0xdaryl — COMMENTED** [review]
Thanks for the pull request.  Are all these changes to fix one problem, or are there multiple fixes rolled into the single commit?  If there are multiple problems then it might be easier to absorb this (as a reviewer at least) if they were separated into different commits.

**Comment by kevindean12** [comment]
Thanks for taking a look at it! It is a single problem that can manifest in several places - but, agreed, the way the diff picks up these changes makes it pretty hard to see what's going on. I'll see if separating the introduction of each of the helper methods into its own commit will at least make the diff clearer.

**Review by 0xdaryl — CHANGES_REQUESTED** [review]
Thanks for breaking into separate commits.

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 3]
```diff
@@ -13112,91 +13103,76 @@ TR::Node *ificmpeqSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier
...
+
    if (removeIfToFollowingBlock(node, block, s) == NULL)
       return NULL;
 
    TR::Node * firstChild = node->getFirstChild(), * secondChild = node->getSecondChild();
 
    if (firstChild == secondChild)
       {
-      s->conditionalToUnconditional(node, block, false);
+      s->conditionalToUnconditional(node, block, !opChangedToCmpNE);
```
[the remark this answers is not in the document]
Comment (kevindean12) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: Good point - changed it back to `true`.

**Comment by 0xdaryl** [comment]
Jenkins test all

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by kevindean12** [comment]
On the win_x86-64 CI build:
```10:47:08  20: [----------] 12 tests from PriorityInterrupt
10:47:11  20: Failed:
10:47:11  20: 	wait returned J9THREAD_INTERRUPTED (2), expected 0 (zero) or J9THREAD_PRIORITY_INTERRUPTED
10:47:11  20: 
10:47:11  20: c:\omr\workspace\pullrequest-win_x86-64\build\fvtest\threadtest\priorityinterrupttest.cpp(286): error: Value of: isRetvalOk(waiter, 0, J9THREAD_PRIORITY_INTERRUPTED)
10:47:11  20:   Actual: false
10:47:11  20: Expected: true
10:47:11  20: [  FAILED  ] PriorityInterrupt.testAllIntr (409 ms)
```

On x86-64 macOS (this and similar repeated):
```
2: ======================================================================
2: ERROR: test_to_opaque_cast_2 (test.cppgentests.CppGeneratorTest.test_to_opaque_cast_2)
2: ----------------------------------------------------------------------
2: Traceback (most recent call last):
2:   File "/Users/runner/work/1/s/jitbuilder/apigen/test/cppgentests.py", line 128, in test_to_opaque_cast_2
2:     self.assertRegexpMatches(self.generator.to_opaque_cast("bar", class_desc),
2:     ^^^^^^^^^^^^^^^^^^^^^^^^
2: AttributeError: 'CppGeneratorTest' object has no attribute 'assertRegexpMatches'
2: 
2: ----------------------------------------------------------------------
2: Ran 146 tests in 0.623s
2: 
2: FAILED (errors=25, skipped=3)
2: warning: The package 'jsonschema' is not installed so certain tests will be skipped
 8/49 Test  #2: TestJitBuilderAPIGenerator ..........***Failed    1.35 sec
```

And what I think is a familiar unrelated failure on macOS, but not positive it's the same as what's been seen elsewhere:
```
34: /Users/runner/work/1/s/fvtest/porttest/omrsockTest.cpp:1259: Failure
34:       Expected: privateOmrPortLibrary->sock_accept(privateOmrPortLibrary, serverSocket, &connectedServerSockAddr, &connectedServerSocket)
34:       Which is: -20
34: To be equal to: 0
34: [  FAILED  ] PortSockTest.poll_functionality_many_sockets (1 ms)
34: [----------] 19 tests from PortSockTest (1007 ms total)
34: 
34: [==========] 246 tests from 21 test cases ran. (193883 ms total)
34: [  PASSED  ] 244 tests.
34: [  FAILED  ] 2 tests, listed below:
34: [  FAILED  ] PortSockTest.poll_functionality_basic
34: [  FAILED  ] PortSockTest.poll_functionality_many_sockets
```

@0xdaryl do you know if these are new failures on this branch, or something seen elsewhere?

**Review by 0xdaryl — APPROVED** [review]
I don't think any of these failures are related to the changes in this PR.  Merging.

**Pull request merged**